platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/LongList.java

```diff
@@ -39,21 +49,70 @@
+
+    static {
+        try {
+            final Field f = Unsafe.class.getDeclaredField("theUnsafe");
+            f.setAccessible(true);
+            UNSAFE = (Unsafe) f.get(null);
+        } catch (final NoSuchFieldException | SecurityException | IllegalArgumentException | IllegalAccessException e) {
+            throw new InternalError(e);
+        }
+    }
+
+    private static final Logger logger = LogManager.getLogger(LongList.class);
     /** A suitable default for the maximum number of longs that may be stored (32GB of longs). */
     protected static final long DEFAULT_MAX_LONGS_TO_STORE = 4_000_000_000L;
     /** The maximum number of longs to store per chunk. */
     protected static final int MAX_NUM_LONGS_PER_CHUNK = Math.toIntExact(16_000L * (MEBIBYTES_TO_BYTES / Long.BYTES));
     /** A suitable default for the number of longs to store per chunk. */
     protected static final int DEFAULT_NUM_LONGS_PER_CHUNK = Math.toIntExact(8L * (MEBIBYTES_TO_BYTES / Long.BYTES));
+
+    /** A suitable default for the reserved buffer length that the list should have before minimal
+     * index in the list
+     */
+    public static final int DEFAULT_RESERVED_BUFFER_LENGTH = Math.toIntExact(2L * MEBIBYTES_TO_BYTES / Long.BYTES);
+
+    /** Maximum number of chunks allowed.*/
+    public static final int MAX_NUM_CHUNKS = 2 << 14;
+
+    /**Initial file format*/
+    private static final int INITIAL_VERSION = 1;
+    /** File format that supports min valid index */
+    private static final int MIN_VALID_INDEX_SUPPORT_VERSION = 2;
     /** The version number for format of current data files */
-    private static final int FILE_FORMAT_VERSION = 1;
-    /** The number of bytes to read for header */
-    protected static final int FILE_HEADER_SIZE = Integer.BYTES + Integer.BYTES + Long.BYTES + Long.BYTES;
+    private static final int CURRENT_FILE_FORMAT_VERSION = MIN_VALID_INDEX_SUPPORT_VERSION;
+    /** The number of bytes required to store file version */
+    protected static final int VERSION_METADATA_SIZE = Integer.BYTES;
+    /** The number of bytes to read for format metadata, v1: <br>
+     * - number of longs per chunk<br>
+     * - max index that can be stored<br>
+     * - max number of longs supported by the list<br>
+     */
+    protected static final int FORMAT_METADATA_SIZE_V1 = Integer.BYTES + Long.BYTES + Long.BYTES;
+    /** The number of bytes to read for format metadata, v2:
+     * - number of longs per chunk<br>
+     * - max number of longs supported by the list<br>
+     * - min valid index<br>
+     */
+    protected static final int FORMAT_METADATA_SIZE_V2 = Integer.BYTES + Long.BYTES + Long.BYTES;
+    /** The number for bytes to read for file header, v1 */
+    protected static final int FILE_HEADER_SIZE_V1 = VERSION_METADATA_SIZE + FORMAT_METADATA_SIZE_V1;
+    /** The number for bytes to read for file header, v2 */
+    protected static final int FILE_HEADER_SIZE_V2 = VERSION_METADATA_SIZE + FORMAT_METADATA_SIZE_V2;
```

Comment:
Just a thought... There are so many static final fields related to versions. What if to group them one way or another? For example:
```
private static class VERSION_ORIGINAL {
    public static final int VERSION_NUM = ...;
    public static final int METADATA_SIZE = ...;
    public static final int FILE_HEADER_SIZE = ...;
}

private static class VERSION_MIN_VALID_INDEX {
    ...
}
```

I don't know if it would be more or less readable, though. What do you think about it?

Reply:
I tried it out and it looks less readable after the refactoring:
```
                if (formatVersion == VersionOriginal.VERSION) {
                    formatMetadataSize = VersionOriginal.FORMAT_METADATA_SIZE;
                    currentFileHeaderSize = VersionOriginal.FILE_HEADER_SIZE;
                } else if (formatVersion == VersionMinValidIndexSupport.VERSION) {
                    formatMetadataSize = VersionMinValidIndexSupport.FORMAT_METADATA_SIZE;
                    currentFileHeaderSize = VersionMinValidIndexSupport.FILE_HEADER_SIZE;
                }
```

vs                 
```
               if (formatVersion == INITIAL_VERSION) {
                    formatMetadataSize = FORMAT_METADATA_SIZE_V1;
                    currentFileHeaderSize = FILE_HEADER_SIZE_V1;
                } else if (formatVersion == MIN_VALID_INDEX_SUPPORT_VERSION) {
                    formatMetadataSize = FORMAT_METADATA_SIZE_V2;
                    currentFileHeaderSize = FILE_HEADER_SIZE_V2;
                }    
```

So, I don't think it worth changing this way.